x/gov/keeper/proposal.go

```diff
@@ -7,39 +7,70 @@ import (
 	sdk "github.com/cosmos/cosmos-sdk/types"
 	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
 	"github.com/cosmos/cosmos-sdk/x/gov/types"
-	"github.com/cosmos/cosmos-sdk/x/gov/types/v1beta1"
+	"github.com/cosmos/cosmos-sdk/x/gov/types/v1beta2"
 )
 
-// SubmitProposal create new proposal given a content
-func (keeper Keeper) SubmitProposal(ctx sdk.Context, content v1beta1.Content) (v1beta1.Proposal, error) {
-	if !keeper.router.HasRoute(content.ProposalRoute()) {
-		return v1beta1.Proposal{}, sdkerrors.Wrap(types.ErrNoProposalHandlerExists, content.ProposalRoute())
-	}
+// SubmitProposal create new proposal given an array of messages
+func (keeper Keeper) SubmitProposal(ctx sdk.Context, messages []sdk.Msg) (v1beta2.Proposal, error) {
+	// Will hold a comma-separated string of all Msg type URLs.
+	msgsStr := ""
+
+	// Loop through all messages and confirm that each has a handler and the gov module account
+	// as the only signer
+	for _, msg := range messages {
+		msgsStr += fmt.Sprintf(",%s", sdk.MsgTypeURL(msg))
+
+		// perform a basic validation of the message
+		if err := msg.ValidateBasic(); err != nil {
+			return v1beta2.Proposal{}, sdkerrors.Wrap(types.ErrInvalidProposalMsg, err.Error())
+		}
+
+		signers := msg.GetSigners()
+		if len(signers) != 1 {
+			return v1beta2.Proposal{}, types.ErrInvalidSigner
+		}
+
+		// assert that the governance module account is the only signer of the messages
+		if !signers[0].Equals(keeper.GetGovernanceAccount(ctx).GetAddress()) {
+			return v1beta2.Proposal{}, sdkerrors.Wrap(types.ErrInvalidSigner, signers[0].String())
+		}
+
+		// for all other message types use the msg service router to see that there is a valid route for that
+		// message.
+		handler := keeper.router.Handler(msg)
+		if handler == nil {
+			return v1beta2.Proposal{}, sdkerrors.Wrap(types.ErrUnroutableProposalMsg, sdk.MsgTypeURL(msg))
+		}
+
+		// Only if it's a MsgExecLegacyContent do we try to execute the
+		// proposal in a cached context.
+		// For other Msgs, we do not verify the proposal messages any further.
+		// They may fail upon execution.
+		// ref: https://github.com/cosmos/cosmos-sdk/pull/10868#discussion_r784872842
+		if msg, ok := msg.(*v1beta2.MsgExecLegacyContent); ok {
+			cacheCtx, _ := ctx.CacheContext()
+			if _, err := handler(cacheCtx, msg); err != nil {
+				return v1beta2.Proposal{}, sdkerrors.Wrap(types.ErrNoProposalHandlerExists, err.Error())
+			}
+		}
```

Comment:
We no longer need to do this right because we have the new MsgServer method for executing legacy content messages.

Reply:
this is related to https://github.com/cosmos/cosmos-sdk/pull/10868#discussion_r784872842.

The current code runs the Msg iif it's a MsgExecLegacyContent (to match old behavior). We could also remove this code, in which case the legacy content will be run for the 1st time in EndBlocker.

I feel that mock running legacy content is safe, those old gov handlers are well-known and don't do any side-effects such as sending IBC messages or writing to disk

Reply:
Ok yeah that's fine with me